indigo/analysis/refs/base.py

```diff
@@ -123,3 +123,150 @@ class RefsFinderENG(BaseRefsFinder):
+
+
+class BaseSectionRefsFinder(LocaleBasedMatcher):
+    """ Finds internal references to sections in documents.
+
+    """
+
+    section_re = None
+    """ This must be defined by a subclass. It should be a compiled regular
+    expression, with named captures for `ref` and `num`.
+    """
+    # TODO: add subsection reference
+    candidate_xpath = None  # this must be defined by a subclass
+
+    # the ancestor elements that can contain references
+    ancestors = ['body', 'mainBody', 'conclusions']
+    # TODO: check on ancestor elements
+
+    def find_references_in_document(self, document):
+        """ Find references in +document+, which is an Indigo Document object.
+        """
+        # we need to use etree, not objectify, so we can't use document.doc.root,
+        # we have to re-parse it
+        root = etree.fromstring(document.content)
+        self.setup(root)
+        self.find_references(root)
+        document.content = etree.tostring(root, encoding='utf-8').decode('utf-8')
+
+    def setup(self, root):
+        self.ns = root.nsmap[None]
+        self.nsmap = {'a': self.ns}
+        self.ref_tag = f'{{{self.ns}}}ref'
+        self.ancestor_xpath = etree.XPath('|'.join(f'.//a:{a}' for a in self.ancestors), namespaces=self.nsmap)
+        self.candidate_xpath = etree.XPath(self.candidate_xpath, namespaces=self.nsmap)
+
+    def make_href(self, node, match):
+        """ Turn this match into an href if a reference is found
+        """
+        num = match.group('num')
+        candidate_elements = node.xpath(f"//a:section[a:num[text()='{num}.']]", namespaces=self.nsmap)
+        if candidate_elements:
+            element_id = candidate_elements[0].get('id')
+            return f'#{element_id}'
+
+    def find_references(self, root):
+        for ancestor in self.ancestor_nodes(root):
+            for candidate in self.candidate_nodes(ancestor):
+                node = candidate.getparent()
+
+                if not candidate.is_tail:
+                    # text directly inside a node
+                    matches = self.section_re.finditer(node.text)
+                    # mark the reference and continue to check the new tail
+                    node = self.mark_refs_in_matches(matches, node, in_tail=False)
+
+                while node is not None and node.tail:
+                    matches = list(self.section_re.finditer(node.tail))
+
+                    # if nothing in the tail matches anymore, we're done
+                    if not matches:
+                        break
+
+                    # there might be just one match that's a baddie,
+                    # in which case we're also done
+                    if len(matches) == 1 and not self.is_valid(matches[0], node):
+                        break
+
+                    # mark the reference and continue to check the new tail
+                    node = self.mark_refs_in_matches(matches, node, in_tail=True)
```

Comment:
This will loop infinitely if the only matches are bad ones. We need a way for this method to tell us that there were no valid matches and we need to stop, perhaps by returning `None`.

Reply:
It may be easier to do the loop here, not in `mark_refs_in_matches`, because iterating over `matches` here is different to iterating over `matches` in the non-tail block above.

Reply:
hmm, but if the first match is bad we don't want to discard the node, which is why we introduced the loop in the first place. this part of it should in theory be the same in both cases: find the first valid reference, mark it and put the rest into the tail, then check the tail.

Reply:
Perhaps this is easier to understand:

```python
if not candidate.is_tail:
    # text directly inside a node
    for match in matches = self.section_re.finditer(node.text):
        if self.is_valid(match, node):
            # mark the reference and continue to check the new tail
            node = self.mark_reference(node, match, False)
            break

while node is not None and node.tail:
    for match in self.section_re.finditer(node.tail):
        if self.is_valid(match, node):
            node = self.mark_reference(node, match, True)
            # the node (and its text) have changed, start again on the tail of the new node
            break
    else:
        # we didn't break out of the loop, so there are no valid matches, give up
        node = None
```

Rather than checking for all the negative cases up front, we handle each match one at a time.

Reply:
My motivation for not having `mark_refs_in_matches` is that (a) it doesn't hide anything complex (it's just looping over an array) and (b) it loses the context necessary to know what to do if there *isn't* a match.

Reply:
gotcha, this makes total sense thanks